tools/tokenize/src/main/scala/org/allenai/nlpstack/tokenize/Ai2Tokenizer.scala

```diff
@@ -0,0 +1,217 @@
+  var averageTokenLength = 6 // low estimates are better
+
+  private object CharacterClass extends Enumeration {
+    type CharacterClass = Value
+    val Whitespace, Punctuation, Dash, Number, Other = Value
+  }
+  import CharacterClass._
+  private val characterClassMap = Map(
+    ' ' -> Whitespace,
+    '\n' -> Whitespace,
+    '\r' -> Whitespace,
+    '\t' -> Whitespace,
+    '\u2028' -> Whitespace,
+    '\u00A0' -> Whitespace,
+    '.' -> Punctuation,
+    ';' -> Punctuation,
+    ':' -> Punctuation,
+    '+' -> Punctuation,
+    '{' -> Punctuation,
+    '}' -> Punctuation,
+    '[' -> Punctuation,
+    ']' -> Punctuation,
+    '(' -> Punctuation,
+    ')' -> Punctuation,
+    '*' -> Punctuation,
+    '^' -> Punctuation,
+    '%' -> Punctuation,
+    '$' -> Punctuation,
+    '#' -> Punctuation,
+    '@' -> Punctuation,
+    '!' -> Punctuation,
+    '~' -> Punctuation,
+    '`' -> Punctuation,
+    ',' -> Punctuation,
+    '\'' -> Punctuation,
+    '"' -> Punctuation,
+    '—' -> Punctuation, // long dash
+    '–' -> Punctuation, // some other dash
+    '’' -> Punctuation,
+    '…' -> Punctuation,
+    '“' -> Punctuation,
+    '”' -> Punctuation,
+    '/' -> Punctuation,
+    '-' -> Dash, // normal dash
+    '?' -> Punctuation,
+    '<' -> Punctuation,
+    '>' -> Punctuation,
+    '‘' -> Punctuation,
+    '•' -> Punctuation,
+    '&' -> Punctuation,
+    '‰' -> Punctuation,
+    '|' -> Punctuation,
+    '=' -> Punctuation,
+    '~' -> Punctuation,
+    '\\' -> Punctuation,
+    '°' -> Punctuation
+  ).withDefaultValue(Other)
+
+  // The code supports replacing one sequence with another (as long as the replacement sequence is
+  // as long or shorter than the pattern sequence). In this version, the replacement sequence is
+  // always of length 1, but hard-coding that assumption has no bearing on performance, and it
+  // doesn't make the code shorter, so I left it this way.
+  private val specialCases = IndexedSeq(
+    IndexedSeq("'", "s") -> "'s", // He's doing the thing.
+    IndexedSeq("'", "re") -> "'re", // You're doing the thing.
+    IndexedSeq("'", "ll") -> "'ll", // I'll do the thing.
+    IndexedSeq("'", "m") -> "'m", // I'm doing the thing.
+    IndexedSeq("'", "d") -> "'d", // You'd do the same thing.
+    IndexedSeq("'", "t") -> "'t", // I can't do the thing.
+    IndexedSeq("'", "ve") -> "'ve", // I should've done the thing.
+    IndexedSeq("e", ".", "g", ".") -> "e.g.",
+    IndexedSeq("i", ".", "e", ".") -> "i.e.",
+    IndexedSeq("a", ".", "m", ".") -> "a.m.",
+    IndexedSeq("p", ".", "m", ".") -> "p.m.",
+    IndexedSeq("et", ".", "al", ".") -> "et.al.",
+    IndexedSeq("ph", ".", "d", ".") -> "ph.d.",
+    IndexedSeq("mr", ".") -> "mr.",
+    IndexedSeq("ms", ".") -> "ms.",
+    IndexedSeq("m", ".", "sc", ".") -> "m.sc.",
+    IndexedSeq("b", ".", "sc", ".") -> "b.sc.",
+    IndexedSeq("b", ".", "a", ".") -> "b.a.",
+    IndexedSeq("etc", ".") -> "etc.",
+    IndexedSeq("u", ".", "s", ".") -> "u.s."
```

Comment:
Partially from comparing with Factorie, and partially from an informal discussion with @pdturney and @tusharkhot. I expect we'll have to update this from time to time.

The real problem is that the time it takes scales roughly linear with the length of this list. If it gets too long, I'll have to come up with some better rules to scope it down.